### ISSUE TITLE
Batches and Promises in the Client

### ISSUE THREAD

**Issue by ardenpm** [tedeh/jayson#96]
It does not seem to be possible to combine batches with promises at the moment. Creating an array of requests using the promise subtree client results in an array of promises which start executing immediately and so not in the same request. Since batches and promises both use the method of not specifying a callback it does not seem possible to distinguish between them.

Not sure what the best solution to this might be, right now I create two clients, a promise one and a non-promise one and build the batch up with the non-promise one then execute it with the promise based one. This works but makes for confusing code. Hopefully I am missing something obvious but would be good to have a way to use batches with the promise version of the client.

**Comment by Overtorment** [comment]
Could not understand the issue, but just for any case, I'm using Promise.all()
```
function getreceivedbyaddress (address) {
  let reqs = [
    client.request('getreceivedbyaddress', [address, 0]),
    client.request('getreceivedbyaddress', [address, 6])
  ]

  return Promise.all(reqs)
}
```

Hope it helps. There are also tricks to launch promises sequentially. Cheers!

**Comment by ardenpm** [comment]
Just to elaborate a bit further, the issue is that in the example you show, two requests will actually be made to the server, not a single request with a JSON-RPC 2.0 batch. At least that is the behaviour I am seeing. This is fine if you just want to do multiple commands and don't care about them being in a batch.

However there are very specific use cases where you want to use a batch and this must be done in a single request per the JSON-RPC 2.0 spec. For example, if the first command in the batch revives some state which the subsequent commands operate on.

In our JSON-RPC server this is often the case since it is backed by a transactional system and all operations within a single request happen in the same transaction. The only way to do multiple things in one request and therefore the same transaction is to use the JSON-RPC batch mechanism. I can imagine this is a fairly common use case.

**Comment by tedeh** [comment]
Solved in `v2.0.5`.

You can see the updated README for an example: https://github.com/tedeh/jayson#promise-batches
